### PR TITLE
many: change "transactional" flag to a "transaction" option

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Change the --transactional flag to an option called --transaction with
two possible values, "per-snap" and "all-snaps". The former is the
default, and makes updates/installs transactional per snap, while the
latter makes the API call transactional on the specified set of snaps,
so if one fails all the action is reverted.

### CODE REVIEW

File: cmd/snap/cmd_snap_op.go [hunk 1]
```diff
@@ -1146,7 +1146,7 @@ func init() {
 			// TRANSLATORS: This should not start with a lowercase letter.
 			"ignore-running": i18n.G("Ignore running hooks or applications blocking the installation"),
 			// TRANSLATORS: This should not start with a lowercase letter.
-			"transactional": i18n.G("Install a set of snaps transactionally."),
+			"transaction": i18n.G("Create one transaction per-snap (default), or create one transaction for all the specified snaps"),
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:1149]: s/Create/Have/

File: cmd/snap/cmd_snap_op.go [hunk 2]
```diff
@@ -1167,7 +1167,7 @@ func init() {
 			// TRANSLATORS: This should not start with a lowercase letter.
 			"leave-cohort": i18n.G("Refresh the snap out of its cohort"),
 			// TRANSLATORS: This should not start with a lowercase letter.
-			"transactional": i18n.G("Refresh a set of snaps transactionally."),
+			"transaction": i18n.G("Create one transaction per-snap (default), or create one transaction for all the specified snaps"),
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:1170]: s/Create/Have/

File: daemon/api_sideload_n_try.go [hunk 3]
```diff
@@ -157,7 +157,11 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
 	flags.RemoveSnapPath = true
 	flags.Unaliased = isTrue(form, "unaliased")
 	flags.IgnoreRunning = isTrue(form, "ignore-running")
-	flags.Transactional = isTrue(form, "transactional")
+	trasactionVals := form.Values["transaction"]
+	flags.Transaction = client.TransactionPerSnap
+	if len(trasactionVals) > 0 {
+		flags.Transaction = client.TransactionType(trasactionVals[0])
```
Comment (pedronis) [inline daemon/api_sideload_n_try.go:163]: this code and snapInstruction.validate should check that this has actually one of the 2 permitted values or it's indeed empty

**Review by pedronis — COMMENTED** [review]
thanks for this, overall looks good, some small comments

**Comment by alfonsosanchezbeato** [comment]
@pedronis thanks for your review, I've done the requested changes now.

File: daemon/api_sideload_n_try.go [hunk 4]
```diff
@@ -160,7 +160,13 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
 	trasactionVals := form.Values["transaction"]
 	flags.Transaction = client.TransactionPerSnap
 	if len(trasactionVals) > 0 {
-		flags.Transaction = client.TransactionType(trasactionVals[0])
+		switch trasactionVals[0] {
+		case string(client.TransactionPerSnap), string(client.TransactionAllSnaps):
+			flags.Transaction = client.TransactionType(trasactionVals[0])
+		default:
+			return BadRequest(`transaction must be either "%s" or "%s"`,
```
Comment (pedronis) [inline daemon/api_sideload_n_try.go:167]: you could use %q instead of "%s"

File: daemon/api_snaps.go [hunk 5]
```diff
@@ -258,6 +258,15 @@ func (inst *snapInstruction) validate() error {
 			}
 		}
 	}
+	switch inst.Transaction {
+	case "":
+	case client.TransactionPerSnap, client.TransactionAllSnaps:
+		if inst.Action != "install" && inst.Action != "refresh" {
+			return fmt.Errorf(`transaction type is invalid for "%s" actions`, inst.Action)
```
Comment (pedronis) [inline daemon/api_snaps.go:265]: transaction type is unsupported for %q action:

**Review by pedronis — APPROVED** [review]
two small nitpicks

File: daemon/api_sideload_n_try_test.go [hunk 6]
```diff
@@ -1093,3 +1093,30 @@ func (s *trySuite) TestTryChangeConflict(c *check.C) {
 	rspe := daemon.TrySnap(st, tryDir, snapstate.Flags{}).(*daemon.APIError)
 	c.Check(rspe.Kind, check.Equals, client.ErrorKindSnapChangeConflict)
 }
+
+func (s *sideloadSuite) TestSideloadSnapInvalidTransaction(c *check.C) {
+	s.daemon(c)
+
+	// try a multipart/form-data upload with missing "name"
```
Comment (MiguelPires) [inline daemon/api_sideload_n_try_test.go:1100]: I don't think this is relevant to this test, probably left over from `TestSideloadSnapNotValidFormFile`

File: daemon/api_sideload_n_try_test.go [hunk 7]
```diff
@@ -1093,3 +1093,30 @@ func (s *trySuite) TestTryChangeConflict(c *check.C) {
...
+	content := "" +
+		"----hello--\r\n" +
+		"Content-Disposition: form-data; filename=\"x\"\r\n" +
+		"\r\n" +
+		"xyzzy\r\n" +
+		"----hello--\r\n"
+	content += "Content-Disposition: form-data; name=\"transaction\"\r\n" +
+		"\r\n" +
+		"xyz\r\n" +
+		"----hello--\r\n"
```
Comment (MiguelPires) [inline daemon/api_sideload_n_try_test.go:1110]: Actually, I think this whole thing can be simplified a bit by removing the form part with the missing name
```suggestion
	content := "" +
		"----hello--\r\n" +
		"Content-Disposition: form-data; name=\"transaction\"\r\n" +
		"\r\n" +
		"xyz\r\n" +
		"----hello--\r\n"
```

**Review by MiguelPires — APPROVED** [review]
LGTM. Left a couple of nitpick comments and some suggestions for improvements in the cmd docs and tests

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the comments, I've addressed them now.

**Pull request merged**